File: evaluation/results/scenario_1a/pymc/llama3.1:8b/run_03/extracted_code.py
# Import necessary libraries
import pandas as pd
import pymc as pm
import numpy as np
import matplotlib.pyplot as plt

# Load data from CSV file
data = pd.read_csv('cases.csv')

# Convert date to datetime format for easier manipulation
data['date'] = pd.to_datetime(data['date'])

# Create a time-series index and align it with the data
ts_index = pd.date_range(start=data['date'].min(), end=data['date'].max())
aligned_data = data.set_index('date').reindex(ts_index).fillna(0)  # Fill missing values with 0

# Define the model
def rt_model(d, n_cases, tau=1/3.8):  # Using a typical estimate for tau (mean generation time)
    mu = pm.Normal('mu', mu=n_cases, sigma=10, shape=len(n_cases))
    beta = pm.Deterministic('beta', 1 / (d + tau))  # Deterministic for computational efficiency
    rt = pm.Deterministic('rt', n_cases * beta)  
    return rt

# Fit the model using PyMC
with pm.Model() as model:
    d = aligned_data.index.to_numpy()
    n_cases = aligned_data['cases'].values
    rt_samples = pm.sample_posterior_predictive(rt_model(d, n_cases), samples=10000)

# Extract Rt estimates for each time point
rt_estimates = np.mean(rt_samples['rt'], axis=0)

# Plot Rt over time (optional)
plt.figure(figsize=(8,6))
plt.plot(aligned_data.index, rt_estimates, color='blue')
plt.fill_between(aligned_data.index, 0, rt_estimates, alpha=0.2, color='blue')
plt.xlabel('Date')
plt.ylabel('Rt')
plt.title('Time-Varying Reproduction Number (Rt) Estimate')
plt.show()

# Print summary statistics for Rt estimates
print(f"Mean Rt estimate: {np.mean(rt_estimates):.4f}")
print(f"Median Rt estimate: {np.median(rt_estimates):.4f}")

File: evaluation/results/scenario_1b/stan/llama3.1:8b/run_03/extracted_code.R
# Load libraries
library(cmdstanr)
library(tidyverse)

# Load case counts from CSV
cases <- read_csv("cases.csv")

# Define Stan model for renewal equation (model.stan)
cat(read_file("renewal_model.stan"), file = "renewal_model.stan")

# Fit the model using cmdstanr
fit <- stan(
  file = "renewal_model.stan",
  data = list(n_cases = nrow(cases), 
              cases = cases$cases, 
              dates = as.numeric(as.Date(cases$date)), 
              S = 7, # 7-day generation interval distribution 
              T = 365) # Time horizon
)

# Extract Rt estimates for each time point
rt_estimates <- data.frame(date = fit$summary$Rt)

# Create a plot of Rt over time ( example )
ggplot(rt_estimates, aes(x = date)) + 
  geom_line() + 
  theme_classic()

